pkg/messaging/nats/tracing/publisher.go

```diff
@@ -0,0 +1,61 @@
+package tracing
+
+import (
+	"bytes"
+
+	"github.com/mainflux/mainflux/pkg/messaging"
+	"github.com/opentracing/opentracing-go"
+	"github.com/opentracing/opentracing-go/ext"
+)
+
+// traced ops
+const (
+	publishOP = "publish_op"
+)
+
+var _ messaging.Publisher = (*publisherMiddleware)(nil)
+
+type publisherMiddleware struct {
+	publisher messaging.Publisher
+	tracer    opentracing.Tracer
+}
+
+func New(publisher messaging.Publisher, tracer opentracing.Tracer) *publisherMiddleware {
+	return &publisherMiddleware{
+		publisher: publisher,
+		tracer:    tracer,
+	}
+}
+
+// Close implements messaging.Publisher
+func (pm *publisherMiddleware) Close() error {
+	return pm.publisher.Close()
+}
+
+// Publish implements messaging.Publisher
+func (pm *publisherMiddleware) Publish(topic string, msg *messaging.Message) error {
+	var sc opentracing.SpanContext = nil
+	var err error
+	//extract span context from message
+	if len(msg.Span) != 0 {
+		parentSCBuffer := bytes.NewBuffer(msg.Span)
+		sc, err = pm.tracer.Extract(opentracing.Binary, parentSCBuffer)
+
+		if err != nil && err != opentracing.ErrSpanContextNotFound {
+			return err
+		}
+	}
+
+	// start new span
+	span := pm.tracer.StartSpan(publishOP, ext.SpanKindProducer, opentracing.ChildOf(sc))
+	ext.MessageBusDestination.Set(span, msg.Subtopic)
+	defer span.Finish()
+
+	dataBuffer := bytes.NewBuffer([]byte{})
+
+	if err := pm.tracer.Inject(span.Context(), opentracing.Binary, dataBuffer); err != nil {
+		return err
+	}
+	msg.Span = dataBuffer.Bytes()
```

Comment:
This is a good example of code that can be put into a middleware.

Reply:
https://github.com/SammyOina/mainflux/blob/0bfccd5565ffe9ee74b30b2fc25ae4037da8e06c/pkg/messaging/brokers/brokers_nats.go#L27-L31

The code is used as a middleware, or is there a specific pattern to use @dborovcanin

Reply:
Sorry, I missed this. I guess because it only implements Publisher but no Subscriber. This is the correct direction, but we need something similar for Publisher (or rather Message Handler).